server/user.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * @fileoverview user function handler
+ */
+var fs = require('fs');
+var assert = require('assert');
+
+var log = require('./log.js');
+var utils = require('./utils.js');
+var database = require('./database.js');
+
+/** @type {user} */
+module.exports = user;
+
+/**
+ * @constructor
+ */
+function user() {}
+
+/** @enum {string} */
+user.QueryType = {
+  SIGNUP: 'sign-up',
+  SIGNIN: 'sign-in'
+};
+
+/** @const {RegExp} */
+user.VALID_EMAIL_REGEX =
+  /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/;
+
+/** @const {RegExp} */
+user.VALID_USERNAME_REGEX = /^\w{6,}$/;
+
+/** @const {RegExp} */
+user.VALID_PASSWORD_REGEX = /^\w{8,}$/;
+
+/**
+ * Expire time of cookie.
+ * @const {number}
+ */
+user.cookieExpireTime = 24 * 60 * 60 * 1000;
+
+/**
+ * @typedef {{
+ *   email: (string|undefined),
+ *   username: string,
+ *   password: string,
+ *   confirmed: (boolean|undefined)
+ * }}
+ */
+user.Info;
+
+/**
+ * @typedef {{
+ *   error: string
+ * }}
+ */
+user.Error;
+
+/**
+ * Checks the user information for signing in.
+ * @param {!mongodb.Db} db Session database.
+ * @param {!user.Info} userInfo User Information.
+ * @param {function(!Object)} callback Callback function.
+ */
+user.signUp = function(db, userInfo, callback) {
+  var cursor = db.collection('userInfo').find({
+    $or: [{username: userInfo.username}, {email: userInfo.email}]
+  });
+  var data = [];
+  cursor.each(function(err, doc) {
+    assert.equal(err, null, err);
+    if (doc != null) {
+      data.push(doc);
+    } else {
+      var result = user.signupUser(db, data);
+      callback(result);
+    }
+  });
+};
+
+/**
+ * Checks the user information for signing in.
+ * @param {!mongodb.Db} db Session database.
+ * @param {!user.Info} userInfo User Information.
+ * @param {function(!Object)} callback Callback function.
+ */
+user.signIn = function(db, userInfo, callback) {
+  var data = [];
+  var query = {$and: [{username: userInfo.username},
+    {password: userInfo.password}]};
+  database.getOne(db.collection('userInfo'), query, data, function(data) {
+    var result = authenticateCallback(data);
+    callback(result);
+  });
+
+  var authenticateCallback = function(data) {
```

Comment:
Looks like this nested function is sync and short. It is not worthy to be nested and you can just include this error check in the getOne callback above.